pkg/api/mock/mock_api_implementation.go

```diff
@@ -0,0 +1,241 @@
+		for _, o := range pageData {
+			// since we are in a goroutine, we might already be in the next iteration of this loop
+			// at the time the receiving end of this channel calls the closure. Having a loop-body
+			// scoped variables makes the data for the closure perfectly identified.
+			closureData := o
+			c <- func(out types.Object) error {
+				reflect.ValueOf(out).Elem().Set(reflect.ValueOf(closureData).Elem())
+
+				select {
+				case <-ctx.Done():
+				case objectRetrieved <- true:
+				}
+
+				return nil
+			}
+
+			select {
+			case <-ctx.Done():
+				break outer
+			case <-objectRetrieved:
+			}
+		}
+	}
+
+	close(c)
+}
+
+// Create stores an types.Object to the MockAPIs local storage and sets a RandomIdentifier
+// if not already present. No validation is performed!
+func (a mockAPI) Create(ctx context.Context, o types.Object, opts ...types.CreateOption) error {
+	if isTag, err := a.tagOperation(o, tagOperationCreate); isTag {
+		return err
+	}
+
+	identifier := test.RandomIdentifier()
+	v := reflect.Indirect(reflect.ValueOf(o))
+	t := reflect.TypeOf(o).Elem()
+	for i := 0; i < t.NumField(); i++ {
+		if tag, ok := t.Field(i).Tag.Lookup("anxcloud"); ok &&
+			tag == "identifier" &&
+			t.Field(i).Type.Kind() == reflect.String &&
+			v.Field(i).String() == "" {
+			v.Field(i).SetString(identifier)
+		}
+	}
+
+	identifier, err := api.GetObjectIdentifier(o, true)
+	if err != nil {
+		return err
+	}
+
+	cloned, err := copystructure.Copy(o)
+	if err != nil {
+		return err
+	}
+
+	a[identifier] = &mockAPIObject{Object: cloned.(types.Object), Tags: make(map[string]bool)}
+
+	return nil
+}
+
+// Update overwrites a types.Object in MockAPIs local storage
+func (a mockAPI) Update(ctx context.Context, o types.IdentifiedObject, opts ...types.UpdateOption) error {
+	identifier, err := api.GetObjectIdentifier(o, true)
+	if err != nil {
+		return err
+	}
+
+	if val, ok := a[identifier]; ok && !val.Deleted {
+		cloned, err := copystructure.Copy(o)
+		if err != nil {
+			return err
+		}
+		val.Object = cloned.(types.Object)
+	} else {
+		return api.ErrNotFound
+	}
+
+	return nil
+}
+
+// Destroy removes a types.Object from MockAPIs local storage
+func (a mockAPI) Destroy(ctx context.Context, o types.IdentifiedObject, opts ...types.DestroyOption) error {
+	if isTag, err := a.tagOperation(o, tagOperationDestroy); isTag {
+		return err
+	}
+
+	identifier, err := api.GetObjectIdentifier(o, true)
+	if err != nil {
+		return err
+	}
+
+	if val, ok := a[identifier]; ok {
+		val.Deleted = true
+	} else {
+		return api.ErrNotFound
+	}
+
+	return nil
+}
+
+type tagOperation int
+
+const (
+	tagOperationCreate tagOperation = iota
+	tagOperationDestroy
+)
+
+func (a mockAPI) tagOperation(o types.Object, op tagOperation) (bool, error) {
+	rwt, ok := o.(*corev1.ResourceWithTag)
+	if !ok {
+		return false, nil
+	}
+
+	obj, ok := a[rwt.Identifier]
+	if !ok {
+		return true, api.ErrNotFound
+	}
+
+	if op == tagOperationCreate {
+		obj.Tags[rwt.Tag] = true
```

Comment:
I think Create returns a 422 error when the requested tag is already on the resource, but not sure if need to match that behavior - should at least be a documented deviation. add a comment